### PR TITLE
feat(ToolBox\IQFeed): prevent unordered ticks to be processed

### LINKED ISSUE #4649: Unordered data coming from IQFeed throw exception when writing data to disk
#### Expected Behavior

Even if some faulty data come, write data to disk

#### Actual Behavior

Throw exception from WriteMinuteOrSecondOrTick method

#### Potential Solution

Manually filter unordered data before passing to LeanDataWriter

#### Reproducing the Problem

Fetching data from NYSE stocks that returns unordered ticks

#### Checklist

- [X] I have completely filled out this template
- [X] I have confirmed that this issue exists on the current `master` branch
- [X] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

- [X] I have provided detailed steps to reproduce the issue

### PR DESCRIPTION
**Pull request by mathpaquette**
closes #4649

#### Description
After discussing with IQFeed support, they confirmed to me that exceptionally they might return unordered data. In order to prevent unexpected exception, I added some checks before return the data.

#### Related Issue

#4649 

#### Motivation and Context

Unexcepted exception from https://github.com/QuantConnect/Lean/blob/master/ToolBox/LeanDataWriter.cs#L364

#### How Has This Been Tested?

Tested locally without impacted the normal behavior.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [X] My code follows the code style of this project.
- [X] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [ ] I have added tests to cover my changes. 
- [X] All new and existing tests passed.
- [X] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

**Comment by mathpaquette** [comment]
image

**Comment by mathpaquette** [comment]
image

**Comment by mathpaquette** [comment]
I've also experienced that on Interval data as well so adding the protection for all data type.

File: ToolBox/IQFeed/IQFeedFileHistoryProvider.cs [hunk 1]
```diff
@@ -270,6 +270,28 @@ private static string GetHistoryRequestKey(string ticker, DateTime startDate, Da
             return $"{ticker}-{startDate}-{endDate}";
         }
 
+        /// <summary>
+        /// Prevent IQFeed from returning unordered data of all sort. This might happen exceptionally.
+        /// </summary>
+        /// <param name="data"></param>
+        /// <returns></returns>
+        private static IEnumerable<BaseData> FilterUnorderedData(IEnumerable<BaseData> data)
```
Comment (Martin-Molinero) [inline ToolBox/IQFeed/IQFeedFileHistoryProvider.cs:278]: Curious why we filter instead of ordering the data?

**Review by Martin-Molinero — COMMENTED** [review]
Thanks for the fix 👍 
Interesting that this affects **historical** IQFeed data, it seems to me they could order it on their side.

File: ToolBox/IQFeed/IQFeedFileHistoryProvider.cs (hunk 1 above)
Comment (mathpaquette) [reply ToolBox/IQFeed/IQFeedFileHistoryProvider.cs:278]: good question... basically its just unexpected but might happen. I dont what to impact the overperformance for 0.001%

**Comment by mathpaquette** [comment]
> Thanks for the fix 👍
> Interesting that this affects **historical** IQFeed data, it seems to me they could order it on their side.

Yes I agree. Ill follow up with them. But I think we need this little patch for now.

File: ToolBox/IQFeed/IQFeedFileHistoryProvider.cs (hunk 1 above)
Comment (Martin-Molinero) [reply ToolBox/IQFeed/IQFeedFileHistoryProvider.cs:278]: Sounds good 👍

**Review by Martin-Molinero — APPROVED** [review]

**Pull request merged**